### ISSUE TITLE
`sst-env.d.ts` broken in monorepos

### ISSUE THREAD

**Issue by zvictor** [sst/ion#958]
# Description

Moving some files (?) out of root breaks typing.

# Steps to reproduce
```haskell
git clone git@github.com:sst/ion.git
cd ion/examples/cloudflare-trpc
mkdir trpc
mv *.ts ./trpc
cp tsconfig.json trpc/tsconfig.json
```

lastly, update paths in `sst.config.ts`:
```diff
diff --git a/examples/cloudflare-trpc/sst.config.ts b/examples/cloudflare-trpc/sst.config.ts

     const trpc = new sst.cloudflare.Worker("Trpc", {
       url: true,
-      handler: "index.ts",
+      handler: "trpc/index.ts",
     });

     const client = new sst.cloudflare.Worker("Client", {
       url: true,
       link: [trpc],
-      handler: "client.ts",
+      handler: "trpc/client.ts",
     });
```

# Result

Both `trpc/sst-env.d.ts` and `sst-env.d.ts` define the same resources, leading to incompatible types

```haskell
Subsequent property declarations must have the same type.
Property '"Trpc"' must be of type '{ type: "sst.cloudflare.Worker"; url: string; }',
but here has type 'Service<undefined>'.ts(2717)
```

```ts
> sst-env.d.ts

import "sst"
declare module "sst" {
  export interface Resource {
    "Client": {
      "type": "sst.cloudflare.Worker"
      "url": string
    }
    "Trpc": {
      "type": "sst.cloudflare.Worker"
      "url": string
    }
  }
}
export {}
```

```ts
> trpc/sst-env.d.ts

/// <reference path="../sst-env.d.ts" />
// cloudflare 
declare module "sst" {
  export interface Resource {
    "Trpc": import("@cloudflare/workers-types").Service
  }
}
```

**Comment by zvictor** [comment]
Just realized, this a duplicate of #639 and #556.

**Comment by zvictor** [comment]
@jayair how can I help with this task?

**Comment by thdxr** [comment]
so currently we search for `tsconfig.json` to figure out where to generate types. i think we need to change this to search for package.json that contains sst and we'll generate types there<br><br>forgot there's cases where people have multiple tsconfig inside a single package

**Comment by zvictor** [comment]
but then if someone also installs sst in another directory (let's say they want to be able to run something like `sst dev -- next dev` there) the problem will be back.

wouldn't it be more reliable to look for `sst.config.ts`? I don't see any use case for multiple sst.config in the same project.

**Comment by thdxr** [comment]
can you try this snapshot build? 0.0.0-1725423305 believe it might be fixed there

**Comment by thdxr** [comment]
note we're now looking to see if @cloudflare/workers-types is in the package.json to determine if we should generate cloudflare bindings

**Comment by isaac-scarrott** [comment]
Hey @thdxr. I have also been having this issue. After installing snapshot `0.0.0-1725423305` it didn't work at first because I had `@cloudflare/workers-types` in both my root `package.json` and one of my workspace directories `package.json`.

<details>
<summary>This caused SST to generate the following `<project-root-directory>/sst-env.d.ts` file:</summary>

```ts
/* This file is auto-generated by SST. Do not edit. */
/* tslint:disable */
/* eslint-disable */
import "sst"
export {}
import "sst"
declare module "sst" {
  export interface Resource {
    "AblyApiKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "AwsAccessKeyId": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "AwsAccessKeySecret": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "AwsRegion": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ClerkJwtVerificationKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ClerkPublishableKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ClerkSecretKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "CloudfareR2SecretAccessKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "CloudflareAccountId": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "CloudflareR2AccessKeyId": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ContentProcessorQueue": {
      "type": "sst.aws.Queue"
      "url": string
    }
    "HuggingFaceAccessToken": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "LambdaCloudflareWorkerApiToken": {
      "type": "random.index/randomString.RandomString"
      "value": string
    }
    "Main": {
      "type": "sst.cloudflare.Worker"
      "url": string
    }
    "OpenAiApiKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "OpenAiOrgId": {
      "type": "sst.sst.Secret"
      "value": string
    }
  }
}
// cloudflare 
declare module "sst" {
  export interface Resource {
    "ContentAssetBucket": import("@cloudflare/workers-types").R2Bucket
    "Database": import("@cloudflare/workers-types").D1Database
    "StaticAssetBucket": import("@cloudflare/workers-types").R2Bucket
  }
}
```
</details>

After removing  and running `sst dev` again it removed the extra declaration of a module, however did not generate another `sst-env.d.ts` in any of my workspace directories.

<details>
<summary>This is the outputted `<project-root-directory>/sst-env.d.ts` file:</summary>

```ts
/* This file is auto-generated by SST. Do not edit. */
/* tslint:disable */
/* eslint-disable */
import "sst"
export {}
declare module "sst" {
  export interface Resource {
    "AblyApiKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "AwsAccessKeyId": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "AwsAccessKeySecret": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "AwsRegion": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ClerkJwtVerificationKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ClerkPublishableKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ClerkSecretKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "CloudfareR2SecretAccessKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "CloudflareAccountId": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "CloudflareR2AccessKeyId": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "ContentAssetBucket": {
      "type": "sst.cloudflare.Bucket"
    }
    "ContentProcessorQueue": {
      "type": "sst.aws.Queue"
      "url": string
    }
    "Database": {
      "type": "sst.cloudflare.D1"
    }
    "HuggingFaceAccessToken": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "LambdaCloudflareWorkerApiToken": {
      "type": "random.index/randomString.RandomString"
      "value": string
    }
    "Main": {
      "type": "sst.cloudflare.Worker"
      "url": string
    }
    "OpenAiApiKey": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "OpenAiOrgId": {
      "type": "sst.sst.Secret"
      "value": string
    }
    "StaticAssetBucket": {
      "type": "sst.cloudflare.Bucket"
    }
  }
}

```
</details>

<details>
<summary>And this is my root `package.json`:</summary>

```json
{
  "private": true,
  "name": "root",
  "version": "1.0.0",
  "workspaces": [
    "apps/*"
  ],
  "scripts": {
    "build": "turbo run build",
    "dev": "npm run dev:sst & turbo run dev --parallel",
    "dev:sst": "sst dev --stage=$SST_STAGE",
    "remove:sst": "sst remove  --stage=$SST_STAGE",
    "deploy:sst": "sst deploy --stage=$SST_STAGE",
    "lint": "eslint ."
  },
  "engines": {
    "node": "~18",
    "yarn": ">=1.22.0"
  },
  "packageManager": "yarn@1.22.22",
  "overrides": {
    "hono": "4.3.11"
  },
  "devDependencies": {
    "@eslint/compat": "^1.0.3",
    "@eslint/eslintrc": "^3.1.0",
    "@eslint/js": "^9.4.0",
    "@pulumi/cloudflare": "^5.25.0",
    "@tsconfig/node18": "^1.0.1",
    "@typescript-eslint/eslint-plugin": "^7.13.0",
    "eslint": "^9.4.0",
    "eslint-config-next": "^14.2.3",
    "eslint-config-prettier": "^9.1.0",
    "eslint-config-turbo": "^2.0.3",
    "eslint-plugin-import": "^2.27.5",
    "eslint-plugin-prettier": "^5.1.3",
    "eslint-plugin-react": "^7.34.2",
    "eslint-plugin-unicorn": "^53.0.0",
    "prettier": "^3.3.1",
    "sst": "0.0.0-1725423305",
    "turbo": "latest",
    "typescript": "5.3.3"
  }
}
```
</details>

<details>
<summary>And this is the `package.json` where my cloudflare workers code lives:</summary>

```json
{
  "name": "api",
  "version": "1.0.0",
  "type": "module",
  "scripts": {
    "build": "tsc --declaration --emitDeclarationOnly",
    "dev": "tsc --watch --declaration --emitDeclarationOnly",
    "db:generate": "drizzle-kit generate",
    "db:migrate": "node src/migrate.js"
  },
  "dependencies": {
    "@clerk/backend": "^1.2.1",
    "@hono/zod-validator": "0.2.1",
    "@huggingface/inference": "^2.7.0",
    "@trpc/server": "^11.0.0-rc.364",
    "ably": "1.2.50",
    "aws4fetch": "^1.0.19",
    "drizzle-kit": "0.24.2",
    "drizzle-orm": "0.33.0",
    "emittery": "^1.0.3",
    "hono": "4.3.11",
    "openai": "4.33.1",
    "superjson": "^2.2.1",
    "zod": "3.22.4",
    "zod-to-json-schema": "^3.23.1"
  },
  "devDependencies": {
    "@cloudflare/workers-types": "^4.20230904.0",
    "@tsconfig/node18": "1.0.1",
    "@types/node": "18.15.5",
    "wrangler": "^3.51.2"
  }
}

```
</details>

**Comment by isaac-scarrott** [comment]
After reading through the code in [this commit](https://github.com/sst/ion/commit/9461ef86134e7c4af4926d0014279f563451e4be#diff-ff9e30a76ffa689cfbc691f489866413788ff59e9523b5c71c1dd37a90b75f83) it seems like I needed to add an `sst` dependency to my package.json where ever I want an `sst-env.d.ts` file to be generated.

Seems to be working now.

I'm just having another issue with the R2 bucket name, however will open another issue for that.

Thanks!

**Comment by thdxr** [comment]
yeah i might actually loosen this and just do all `package.json` even if it doesn't contain sst - some people just install at root and rely on hoisting<br><br>but glad to see it works

**Comment by fredtcaroli** [comment]
Has this been released yet? I'm trying out sst 3.0.122 and `.sst-env.d.ts` is not being created in my packages.

Each package has a `package.json` with a `sst` dependency. But no `sst-env.d.ts` is added.

Granted, maybe my setup is not usual. I'm putting `sst.config.ts` inside an app in my monorepo, it's not a root level thing.

Kind of like this:

```
.
├── package.json  // <-- Root of the repo
├── function
│   ├── package.json
│   ├── index.ts  // <-- Handler
└── sst-app
    ├── package.json
    └── sst.config.ts  // <-- SST 'home'
```

I can create a function like this:
```ts
    new sst.aws.Function('MyFunction', {
      handler: '../function/src/index.handler',
      link: [bucket],
    });
```
(It doesn't work with absolute paths for some reason, only relative. Should I open an issue for this?)

`function/` never gets a `sst-env.d.ts`.

**Comment by thdxr** [comment]
this hasn't been released yet, working on getting it merged today

**Comment by thdxr** [comment]
looking closer at your setup, it still won't work since we expect `sst.config.ts` to be at the root of your monorepo

**Comment by fredtcaroli** [comment]
Oh I see. Do you plan on supporting it?

I was figuring out how to use turbo to have separate deployable units in the same monorepo, sorta like stacks in CDK.

**Comment by austinm911** [comment]
I'm having issues getting typescript to recognize the R2Bucket type in my monorepo

Root package.json
```ts
{
	"name": "monorepo",
	"dependencies": {
		"sst": "3.0.122"
	},
	"devDependencies": {
		"@biomejs/biome": "^1.8.3",
		"@happy-dom/global-registrator": "^15.7.3",
		"@types/bun": "^1.1.8",
		"ts-essentials": "^10.0.2",
		"turbo": "^2.1.1",
		"@cloudflare/workers-types": "^4.20240903.0",
		"typescript": "^5.6.1-rc"
	},
	"engines": {
		"node": ">=20"
	},
	"packageManager": "bun@1.1.27",
	"workspaces": ["apps/*", "packages/*", "infra/*"]
}
```
root sst.env.d.ts
```ts
/* tslint:disable */
/* eslint-disable */
import "sst"
declare module "sst" {
  export interface Resource {
    "api": {
      "type": "sst.cloudflare.Worker"
      "url": string
    }
    "web": {
      "type": "sst.aws.StaticSite"
      "url": string
    }
    "bucket": {
      "type": "sst.cloudflare.Bucket"
    }
  }
}
export {}


apps/api package.json
```ts
{
	"name": "@acme/api",
	"version": "0.0.0",
	"type": "module",
	},
	"dependencies": {
		"@hono/zod-validator": "^0.2.2",
		"@valterra/core": "workspace:*",
		"@valterra/db": "workspace:*",
		"@workos-inc/node": "^7.26.0",
		"hono": "^4.5.11",
		"nanoid": "^5.0.7"
	},
	"devDependencies": {
		"@cloudflare/workers-types": "^4.20240903.0",
		"@acme/config-typescript": "*",
		"wrangler": "^3.75.0"
	}
}
```

sst.env.d.ts in apps/api
```ts
/* tslint:disable */
/* eslint-disable */
/// <reference path="../../sst-env.d.ts" />
// cloudflare 
declare module "sst" {
  export interface Resource {
    "bucket": import("@cloudflare/workers-types").R2Bucket
  }
}
export {}
```

`apps/api/src/routes/route-abc.routes.ts`
```ts
const uploadedFile = await Resource['bucket'].put
// Property 'put' does not exist on type '{ type: "sst.cloudflare.Bucket"; }'.ts(2339)
```

**Comment by thdxr** [comment]
i believe this is fixed in 3.1.0

**Comment by prabal-rje** [comment]
I think it'll be helpful to have this in the documentation. Specifically, the fact that the user must run `npm install --save-dev @cloudflare/workers-types` in the specific directory they are using a Cloudflare Worker in. This is not the case for AWS SDK AFAIK.

Additionally, binding Cloudflare resources with one another isn't quite working properly as explained [here](https://discord.com/channels/983865673656705025/1084270849206526163/1291968701905502269) on Discord. None of this was seemingly necessary for the examples using Cloudflare.

Cheers,
Prabal
